pkg/controller/mongodb/mongodb_tls.go

```diff
@@ -73,16 +76,50 @@ func (r *ReplicaSetReconciler) validateTLSConfig(mdb mdbv1.MongoDB) (bool, error
 }
 
 // getTLSConfigModification creates a modification function which enables TLS in the automation config.
-// The config is only updated after the certs and keys have been rolled out to all pods.
-// The agent needs these to be in place before the config is updated.
-// Once the config is updated, the agents will gradually enable TLS in accordance with: https://docs.mongodb.com/manual/tutorial/upgrade-cluster-to-ssl/
-func getTLSConfigModification(mdb mdbv1.MongoDB) automationconfig.Modification {
-	if !(mdb.Spec.Security.TLS.Enabled && hasRolledOutTLS(mdb)) {
-		return automationconfig.NOOP()
+// It will also ensure that the combined cert-key secret is created.
+func getTLSConfigModification(getUpdateCreator secret.GetUpdateCreator, mdb mdbv1.MongoDB) (automationconfig.Modification, error) {
+	if !mdb.Spec.Security.TLS.Enabled {
+		return automationconfig.NOOP(), nil
+	}
+
+	err := ensureTLSSecret(getUpdateCreator, mdb)
+	if err != nil {
+		return automationconfig.NOOP(), err
+	}
+
+	// The config is only updated after the certs and keys have been rolled out to all pods.
+	// The agent needs these to be in place before the config is updated.
+	// Once the config is updated, the agents will gradually enable TLS in accordance with: https://docs.mongodb.com/manual/tutorial/upgrade-cluster-to-ssl/
+	if hasRolledOutTLS(mdb) {
+		return tlsConfigModification(mdb), nil
+	} else {
+		return automationconfig.NOOP(), nil
+	}
+}
+
+// ensureTLSSecret will create or update the operator-managed Secret containing
+// the concatenated certificate and key from the user-provided Secret.
+func ensureTLSSecret(getUpdateCreator secret.GetUpdateCreator, mdb mdbv1.MongoDB) error {
+	data, err := secret.ReadStringData(getUpdateCreator, mdb.TLSSecretNamespacedName())
+	if err != nil {
+		return err
 	}
 
+	combined := fmt.Sprintf("%s%s", data[tlsSecretCertName], data[tlsSecretKeyName])
```

Comment:
This is checked earlier in the reconciliation by `validateTLSConfig` so they should always exist. It might be clearer to add that here as well. What do you say?

Reply:
I would vote for adding additional checks, if this function is called from a separate context, the validation would be nice to have.

Reply:
maybe it is `ensureTLSSecret` the one that should call `validateTLSConfig`. For 1 reason in particular: The Secret validated by `validateTLSConfig` is not necessarilly the same we have here, because we are fetching it twice from the API.

We can do:

1. getTLSConfigModification (from reconcile)
2. ensureTLSSecret (from 1)
3. `validateTLSConfig` (from 2), and making this function to validate a Secret that is passed as an argument (not reading it again from inside the function). If that function succeeds, it means that the Secret that was validated, is the same one that will be used next

makes sense?

Reply:
Me and Rodrigo discussed this and came to the conclusion that it's hard to fix with our current design. The problem is that `validateTLSConfig` returns a boolean which is used to control the requeue. If we move the function we would have to bubble this result all the way up to the reconciliation loop. We should keep this in mind when doing the reconciliation refactoring later.

For now, I have gone ahead and added extra checks by using the helper functions from the `secret` package. @chatton is that the kind of validation you had in mind?